File: packages/mobile-native/android/src/main/java/com/fractonica/mobile/FractonicaClientModule.kt
package com.fractonica.mobile

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.asCoroutineDispatcher
import kotlinx.coroutines.cancel
import java.util.concurrent.Executors

class FractonicaClientModule : Module() {
  private val dispatcher = Executors.newSingleThreadExecutor { runnable ->
    Thread(runnable, "fractonica-mobile-client")
  }.asCoroutineDispatcher()
  private val clientScope = CoroutineScope(SupervisorJob() + dispatcher)
  private var coordinator: FractonicaClientCoordinator? = null
  private var destroyed = false

  override fun definition() = ModuleDefinition {
    Name("FractonicaClient")

    AsyncFunction("bridgeStatus") {
      coordinator().bridgeStatus()
    }.runOnQueue(clientScope)

    AsyncFunction("clientStatus") {
      coordinator().status()
    }.runOnQueue(clientScope)

    AsyncFunction("clientListRecords") { options: Map<String, Any?> ->
      coordinator().listRecords(options)
    }.runOnQueue(clientScope)

    AsyncFunction("clientGetRecord") { options: Map<String, Any?> ->
      coordinator().getRecord(options)
    }.runOnQueue(clientScope)

    AsyncFunction("clientCreateRecord") { options: Map<String, Any?> ->
      coordinator().createRecord(options)
    }.runOnQueue(clientScope)

    AsyncFunction("clientClaimPairingInvitation") { options: Map<String, Any?> ->
      coordinator().claimPairingInvitation(options)
    }.runOnQueue(clientScope)

    AsyncFunction("clientAcceptPairingInvitation") { options: Map<String, Any?> ->
      coordinator().acceptPairingInvitation(options)
    }.runOnQueue(clientScope)

    AsyncFunction("clientResetLocalInstallation") { options: Map<String, Any?> ->
      coordinator().resetLocalInstallation(options)
    }.runOnQueue(clientScope)

    OnDestroy {
      val current = takeCoordinatorForDestroy()
      current?.shutdown()
      clientScope.cancel()
      dispatcher.close()
    }
  }

  @Synchronized
  private fun coordinator(): FractonicaClientCoordinator {
    if (destroyed) {
      throw FractonicaClientBridgeException(
        "ERR_FRACTONICA_CLIENT_UNAVAILABLE",
        "The local Fractonica client is unavailable.",
      )
    }
    coordinator?.let { return it }
    val context = appContext.reactContext?.applicationContext
      ?: throw FractonicaClientBridgeException(
        "ERR_FRACTONICA_CLIENT_UNAVAILABLE",
        "The local Fractonica client is unavailable.",
      )
    return FractonicaClientCoordinator(context).also { coordinator = it }
  }

  @Synchronized
  private fun takeCoordinatorForDestroy(): FractonicaClientCoordinator? {
    destroyed = true
    return coordinator.also { coordinator = null }
  }
}

File: packages/mobile-native/ios/FractonicaClientModule.swift
import ExpoModulesCore

public final class FractonicaClientModule: Module {
  private let clientQueue = DispatchQueue(label: "org.fractonica.mobile.client", qos: .userInitiated)
  private let coordinator = FractonicaClientCoordinator()

  public func definition() -> ModuleDefinition {
    Name("FractonicaClient")

    AsyncFunction("bridgeStatus") { () throws -> [String: Any] in
      try self.coordinator.bridgeStatus()
    }.runOnQueue(clientQueue)

    AsyncFunction("clientStatus") { () throws -> [String: Any] in
      try self.coordinator.status()
    }.runOnQueue(clientQueue)

    AsyncFunction("clientListRecords") { (options: [String: Any]) throws -> [[String: Any]] in
      try self.coordinator.listRecords(options: options)
    }.runOnQueue(clientQueue)

    AsyncFunction("clientGetRecord") { (options: [String: Any]) throws -> [String: Any]? in
      try self.coordinator.getRecord(options: options)
    }.runOnQueue(clientQueue)

    AsyncFunction("clientCreateRecord") { (options: [String: Any]) throws -> [String: Any] in
      try self.coordinator.createRecord(options: options)
    }.runOnQueue(clientQueue)

    AsyncFunction("clientClaimPairingInvitation") { (options: [String: Any]) throws -> [String: Any] in
      try self.coordinator.claimPairingInvitation(options: options)
    }.runOnQueue(clientQueue)

    AsyncFunction("clientAcceptPairingInvitation") { (options: [String: Any]) throws -> [String: Any] in
      try self.coordinator.acceptPairingInvitation(options: options)
    }.runOnQueue(clientQueue)

    AsyncFunction("clientResetLocalInstallation") { (options: [String: Any]) throws in
      try self.coordinator.resetLocalInstallation(options: options)
    }.runOnQueue(clientQueue)

    OnDestroy {
      self.clientQueue.async {
        self.coordinator.shutdown()
      }
    }
  }
}
